### PR TITLE
Restart engine when bot exceeds per-move time limit

### LINKED ISSUE #893: Bot doesn't play the move supplied by the engine through UCI
Engine responded with a move, but it wasn't played, resulting in bot losing on time. In logs I see

```
2024-01-21 18:33:36,435 chess.engine (engine.py:1062) DEBUG <UciProtocol (pid=69018)>: << position startpos moves e2e4 a7a5 f1c4 a5a4 g1f3 a8a7 d2d3 a7a5 c1d2
2024-01-21 18:33:36,436 chess.engine (engine.py:1062) DEBUG <UciProtocol (pid=69018)>: << go wtime 866320 btime 852039 winc 14000 binc 14000 depth 5 movetime 2000
2024-01-21 18:33:48,442 chess.engine (engine.py:1062) DEBUG <UciProtocol (pid=69018)>: << stop
2024-01-21 18:33:48,442 chess.engine (engine.py:1088) DEBUG <UciProtocol (pid=69018)>: >> bestmove a5c5
2024-01-21 18:37:24,698 __main__ (lichess-bot.py:370) DEBUG Event: {'type': 'correspondence_ping'}
2024-01-21 18:38:57,491 __mp_main__ (lichess-bot.py:697) DEBUG Game state: {'type': 'chatLine', 'room': 'player', 'username': 'tzar_bomba', 'text': '!help'}
2024-01-21 18:38:57,492 lib.conversation (conversation.py:42) INFO *** https://lichess.org/ysYc4BEr/black [player] tzar_bomba: !help
2024-01-21 18:38:57,492 lib.conversation (conversation.py:83) INFO *** https://lichess.org/ysYc4BEr/black [player] arvyy_bot: Supported commands: !wait (wait a minute for my first move), !name, !howto, !eval, !queue
2024-01-21 18:38:57,493 urllib3.connectionpool (connectionpool.py:292) DEBUG Resetting dropped connection: lichess.org
```

so the a5c5 was communicated, but it was never played, despite bot being responsive for chat commands https://lichess.org/ysYc4BEr . Using latest master commit (8deb03cb6b6e10697ae81ca2900902a3bd8e9db0)

### PR DESCRIPTION
**Pull request by MarkZH**
## Type of pull request:
- [X] Bug fix
- [ ] Feature
- [ ] Other

## Description:

If a user specifies a per-move time limit in their configuration file--e.g., `uci_options: go_commands: movetime: 1000` for 1000 milliseconds per move--and the engine exceeds this time limit by more than 10 seconds, the python-chess library will intervene and cancel the search while raising a `TimeoutError`. Previously, this error would be caught and main `play_game()` loop would carry on without sending a move to lichess.org.

Now, the `TimeoutError` is propagated until it hits the `backoff` decoration of `play_game()`. This shuts down the engine (correctly, now, due to changes in engine_wrapper.py) and restarts it when `backoff` restarts `play_game()`. Even if the engine keeps exhibiting the same buggy behavior, the errors will be displayed on the terminal and log, so the user will know that something is wrong. 

## Related Issues:

This PR closes #893 and should be a better solution to #623 than the previous PR #628.

## Checklist:

- [X] I have read and followed the [contribution guidelines](/CONTRIBUTING.md).
- [X] I have added necessary documentation (if applicable).
- [X] The changes pass all existing tests.

### CODE REVIEW

**Comment by MarkZH** [comment]
To see the changes to test_bot/test_bot.py, hide the whitespace changes. Since the buggy_engine made for the new test can only play a set list of moves, I had to change the `thread_for_test()` to allow changing the opponent to something other than stockfish.

File: test_bot/test_bot.py [hunk 1]
```diff
@@ -1,319 +1,354 @@
...
+        file.write(original_strategies)
+    lichess_bot.logger.info("Finished Testing Homemade")
+    assert win == "1"
+    assert os.path.isfile(os.path.join(CONFIG["pgn_directory"],
+                                       "bo vs b - zzzzzzzz.pgn"))
+
+
+@pytest.mark.timeout(30, method="thread")
+def test_buggy_engine() -> None:
+    """Test lichess-bot with Stockfish (UCI)."""
```
Comment (AttackingOrDefending) [inline test_bot/test_bot.py:326]: The comment should be changed since we don't use stockfish.

**Comment by AttackingOrDefending** [comment]
lichess-bot plays black so it always receieves an odd number of moves, so the engine that waits for 11 seconds is the outside engine and not the one managed by lichess-bot. You can change `buggy_engine.py` to
```
"""An engine that takes too much time to make a move during tests."""

import chess
import time

assert input() == "uci"

print("id name Delay tactic", flush=True)
print("id author MZH", flush=True)
print("uciok", flush=True)

delay_performed = False
just_started = True
scholars_mate = ["a2a3", "e7e5", "a3a4", "f8c5", "a4a5", "d8h4", "a5a6", "h4f2"]

while True:
    command, *remaining = input().split()
    if command == "quit":
        break
    elif command == "isready":
        print("readyok", flush=True)
    elif command == "position":
        spec_type, *remaining = remaining
        assert spec_type == "startpos"
        board = chess.Board()
        if remaining:
            moves_label, *move_list = remaining
            assert moves_label == "moves"
            for move in move_list:
                board.push_uci(move)
        if just_started and len(board.move_stack) > 1:
            delay_performed = True
    elif command == "go":
        move_count = len(board.move_stack)
        if move_count == 3 and not delay_performed:
            print("info string delaying move", flush=True)
            delay_performed = True
            time.sleep(12)
        move = scholars_mate[move_count]
        print(f"bestmove {move}", flush=True)
        just_started = False
```
and line 346 in `test_bot.py` to `CONFIG["engine"]["uci_options"] = {"go_commands": {"movetime": 100}}`.

**Comment by MarkZH** [comment]
I'm going to fix this PR so the git blame for test_bot.py isn't rendered completely useless. I think my editor changed the newlines on every line.

**Comment by MarkZH** [comment]
This PR is replaced by #897 to reduce clutter due to whitespace issues.

**Pull request closed without merging**